include/psa/crypto_values.h

```diff
@@ -2011,4 +2076,175 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
 
 /**@}*/
 
+/** \defgroup pake Password-authenticated key exchange (PAKE)
+ * @{
+ */
+
+/** The first peer in a balanced PAKE.
+ *
+ * Although balanced PAKE protocols are symmetric, some of them needs an
+ * ordering of peers for the transcript calculations. If the protocol does not
+ * need this, either #PSA_PAKE_SIDE_FIRST or #PSA_PAKE_SIDE_SECOND are
+ * accepted.
+ */
+#define PSA_PAKE_SIDE_FIRST                ((psa_pake_side_t)0x0001)
+
+/** The second peer in a balanced PAKE.
+ *
+ * Although balanced PAKE protocols are symmetric, some of them needs an
+ * ordering of peers for the transcript calculations. If the protocol does not
+ * need this, either #PSA_PAKE_SIDE_FIRST or #PSA_PAKE_SIDE_SECOND are
+ * accepted.
+ */
+#define PSA_PAKE_SIDE_SECOND                ((psa_pake_side_t)0x0002)
+
+/** The client in an augmented PAKE.
+ *
+ * Augmented PAKE protocols need to differentiate between client and server.
+ */
+#define PSA_PAKE_SIDE_CLIENT                ((psa_pake_side_t)0x0101)
+
+/** The server in an augmented PAKE.
+ *
+ * Augmented PAKE protocols need to differentiate between client and server.
+ */
+#define PSA_PAKE_SIDE_SERVER                ((psa_pake_side_t)0x0102)
+
+/** The PAKE uses elliptic curves.
+ *
+ * The corresponding family type is ::psa_ecc_family_t. in determining a
+ * specific curve in the family ::psa_pake_bits_t values are interpreted in the
+ * exact same way as ::psa_key_bits_t would.
+ *
+ * Input and output during the operation can involve group elements and scalar
+ * values:
+ * -# The format for group elements is the same as for public keys on the
+ *  specific curve would be. For more information, consult the documentation of
+ *  psa_export_public_key().
+ * -# The format for scalars is the same as for private keys on the specific
+ *  curve would be. For more information, consult the documentation of
+ *  psa_export_key().
+ */
+#define PSA_PAKE_PRIMITIVE_TYPE_CURVE       ((psa_pake_primitive_type_t)0x01)
+
+/** The PAKE uses finite fields based Diffie-Hellman groups.
+ *
+ * The corresponding family type is ::psa_dh_family_t. In determining a
+ * specific group in the family ::psa_pake_bits_t values are interpreted in the
+ * exact same way as ::psa_key_bits_t would.
+ *
+ * Input and output during the operation can involve group elements and scalar
+ * values:
+ * -# The format for group elements is the same as for public keys on the
+ *  specific group would be. For more information, consult the documentation of
+ *  psa_export_public_key().
+ * -# The format for scalars is the same as for private keys on the specific
+ *  group would be. For more information, consult the documentation of
+ *  psa_export_key().
+ */
+#define PSA_PAKE_PRIMITIVE_TYPE_FIELD_DH       ((psa_pake_primitive_type_t)0x02)
+
+/** Construct a PAKE primitive from type, family and bitsize.
```

Comment:
> Why then is it constructed differently?

I am sorry I am not sure I understand, differently from what? Differently from key type? I thought key type does not encode key length.

Reply:
I'm wondering why there is a separate type `psa_pake_primitive_t`, rather than passing a `psa_key_type_t key_agreement_key_type` and a `size_t bits`.

Reply:
Oh I see, one of the reasons is what you guessed, the groups used for key agreement don't cover everything.

For example RSP uses rings instead of groups. The modulus might be similar and technically we might be able to use key agreement key types to represent them, but feels like a hack and might lay the bed for very subtle mistakes.

Even if we look at J-PAKE, we have subtle issues. The key shares might be technically Diffie-Hellman public keys, the ZKP public keys might look like Diffie-Hellman keys and the security of the scheme relies on the Diffie-Hellman assumption (without looking it up I am not sure which one, but my memory is that for key agreement computational is enough, but for the ZKP you need decisional). Even if the security assumptions are aligned, these "keys" are in a complex interaction with each other and I feel uncomfortable using the same type for them as for key agreement.

Another reason is that as far as the interface is concerned this is not a key type. There are no keys of this type at all. As mentioned above, in theory, some of the data exchanged might be considered of this key type, but this is not shown in the interface, because these are not secret values and differentiating data that can be thought of this way in general does come with a cost and doesn't have any benefit.

In summary, technically it would be possible (would potentially introduce new families for key agreement that only would be used in PAKEs), but I think that it is cleaner to separate them.